### PR TITLE
[DUOS-2856][risk=no] List Dataset Performance Update

### PR DESCRIPTION
**Pull request by rushtong**
### Addresses
https://broadworkbench.atlassian.net/browse/DUOS-2856

### Summary
This PR updates the performance of parsing `DataUse` objects when retrieving lists of `Datasets`. We do two things to improve the bottleneck seen here. 
1. Use Gson to do the parsing instead of Jackson's object mapper
2. Use a local cache to avoid reprocessing the same string values into `DataUse` objects

Example dev vs local comparison (using DAC datasets call for DAC 3) is  71512 ms -> 23390 ms which is roughly a 3-fold improvement. More work should still be done to improve performance, however, since a 23 second response is still fairly slow for the volume of data we're processing.

#### Before
Screenshot 2023-12-21 at 3 15 28 PM

#### After
Screenshot 2023-12-21 at 3 32 10 PM

----
Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

- Label PR with a Jira ticket number and include a link to the ticket
- Label PR with a security risk modifier [no, low, medium, high]
- PR describes scope of changes
- Get a minimum of one thumbs worth of review, preferably two if enough team members are available
- Get PO sign-off for all non-trivial UI or workflow changes
- Verify all tests go green
- Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

**Review by fboulnois — APPROVED** [review]
nice work 👍

File: src/main/java/org/broadinstitute/consent/http/db/mapper/DataUseParser.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  public DataUse parseDataUse(String dataUseString) {
+    dataUseCache.computeIfAbsent(dataUseString, s -> {
+      try {
+        return gson.fromJson(dataUseString, DataUse.class);
+      } catch (Exception e) {
+        logWarn(String.format("Unable to parse data use string: '%s'", dataUseString));
+      }
+      return null;
+    });
+    return dataUseCache.get(dataUseString);
```
Comment (pshapiro4broad) [inline src/main/java/org/broadinstitute/consent/http/db/mapper/DataUseParser.java]: `computeIfAbsent()` returns the value so you don't need the cal to `get()` afterwards
```suggestion
    return dataUseCache.computeIfAbsent(dataUseString, s -> {
      try {
        return gson.fromJson(dataUseString, DataUse.class);
      } catch (Exception e) {
        logWarn(String.format("Unable to parse data use string: '%s'", dataUseString));
      }
      return null;
    });
```

File: src/main/java/org/broadinstitute/consent/http/db/mapper/DataUseParser.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import org.broadinstitute.consent.http.models.DataUse;
+import org.broadinstitute.consent.http.util.ConsentLogger;
+import org.broadinstitute.consent.http.util.gson.GsonUtil;
+
+public class DataUseParser implements ConsentLogger {
+
+  private final Gson gson = GsonUtil.gsonBuilderWithAdapters().create();
+  private final Map<String, DataUse> dataUseCache = new HashMap<>();
```
Comment (pshapiro4broad) [inline src/main/java/org/broadinstitute/consent/http/db/mapper/DataUseParser.java:13]: I assume that the lifetime of a jdbi row reducer is only during a single request. If so, then this only improves the cases where the same data use expression occurs in different rows in the result, correct? Vs making the cache global across all operations. That would require using a concurrent map though so maybe it wouldn't be much of an improvement.

File: src/main/java/org/broadinstitute/consent/http/db/mapper/DacWithDatasetsReducer.java [hunk 3]
```diff
@@ -58,7 +60,7 @@ public void accumulate(Map<Integer, Dac> container, RowView rowView) {
 
         if (Objects.nonNull(rowView.getColumn("dataset_data_use", String.class))) {
           String duStr = rowView.getColumn("dataset_data_use", String.class);
-          Optional<DataUse> du = DataUse.parseDataUse(duStr);
+          Optional<DataUse> du = Optional.ofNullable(dataUseParser.parseDataUse(duStr));
           du.ifPresent(dataset::setDataUse);
```
Comment (pshapiro4broad) [inline src/main/java/org/broadinstitute/consent/http/db/mapper/DacWithDatasetsReducer.java]: Is this check necessary? if the result is `null`, what is the harm of calling `setDataUse`?
```
          dataset.setDataUse(dataUseParser.parseDataUse(duStr));
```

File: src/main/java/org/broadinstitute/consent/http/db/mapper/DacWithDatasetsReducer.java [hunk 4]
```diff
@@ -58,7 +60,7 @@ public void accumulate(Map<Integer, Dac> container, RowView rowView) {
 
         if (Objects.nonNull(rowView.getColumn("dataset_data_use", String.class))) {
           String duStr = rowView.getColumn("dataset_data_use", String.class);
```
Comment (pshapiro4broad) [inline src/main/java/org/broadinstitute/consent/http/db/mapper/DacWithDatasetsReducer.java]: Note that this form makes two calls to `rowView.getColumn()`. Storing the result of the first call would avoid this extra fetch.
```suggestion
        String duStr = rowView.getColumn("dataset_data_use", String.class);
        if (duStr != null) {
```

**Review by pshapiro4broad — APPROVED** [review]
Looks OK, I had a few minor comments. I've heard that Jackson is slower than gson but hadn't seen any benchmarks about it; did you do a perf test with Jackson vs gson separate from the cache change? I'm wondering if cached Jackson is that much slower than cached gson.

**Comment by rushtong** [comment]
> I've heard that Jackson is slower than gson but hadn't seen any benchmarks about it; did you do a perf test with Jackson vs gson separate from the cache change? I'm wondering if cached Jackson is that much slower than cached gson.

I didn't do a performance check with Jackson vs Gson but I've seen similar reviews. We originally used Jackson in quite a few parts of consent/ontology codebases and started moving to Gson quite some time ago. I don't have any desire to refactor everything back to Jackson hence the re-write here.

File: src/main/java/org/broadinstitute/consent/http/db/mapper/DataUseParser.java (hunk 1 above)
Comment (rushtong) [reply src/main/java/org/broadinstitute/consent/http/db/mapper/DataUseParser.java]: Thank you, nice suggestion

File: src/main/java/org/broadinstitute/consent/http/db/mapper/DataUseParser.java (hunk 2 above)
Comment (rushtong) [reply src/main/java/org/broadinstitute/consent/http/db/mapper/DataUseParser.java:13]: Yes, mappers/reducers are instantiated per query, so their lifetime is short as is the cache - which I'm fine with for now. We certainly could make this a larger cache refactor but as you mention, we'd need to measure the tradeoffs. What we find in practice is that a lot of datasets have the exact same data use so I think the bulk of the performance gain comes from that fact.

File: src/main/java/org/broadinstitute/consent/http/db/mapper/DacWithDatasetsReducer.java (hunk 3 above)
Comment (rushtong) [reply src/main/java/org/broadinstitute/consent/http/db/mapper/DacWithDatasetsReducer.java]: Great point, I think this was an artifact of the original return signature which was an optional - definitely not necessary any longer.

**Pull request merged**